### PR TITLE
8298614: Support CDS heap dumping for SerialGC and ParallelGC

### PR DESCRIPTION
**Pull request by matias9927**
This patch upstreams [JDK-8333222](https://bugs.openjdk.org/browse/JDK-8333222) from the Leyden repo to support Serial and Parallel GC for CDS. Additional test fixes not in that patch were added. Verified with tier 1-5 tests.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8298614](https://bugs.openjdk.org/browse/JDK-8298614): Support CDS heap dumping for SerialGC and ParallelGC (**Enhancement** - P3)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Ioi Lam](https://openjdk.org/census#iklam) (@iklam - **Reviewer**) Review applies to [905fdfd3](https://git.openjdk.org/jdk/pull/21004/files/905fdfd3d82fcc116fdcbd68dee06b34e681e29c)
 * [Leonid Mesnik](https://openjdk.org/census#lmesnik) (@lmesnik - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21004/head:pull/21004` \
`$ git checkout pull/21004`

Update a local copy of the PR: \
`$ git checkout pull/21004` \
`$ git pull https://git.openjdk.org/jdk.git pull/21004/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21004`

View PR using the GUI difftool: \
`$ git pr show -t 21004`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21004.diff">https://git.openjdk.org/jdk/pull/21004.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21004#issuecomment-2350133236)

### CODE REVIEW

**Comment by dholmes-ora** [comment]
I suggest changing the description (per the upstream issue) to be clear this is for Parallel and serial GC only not all non-G1 collectors.

File: src/hotspot/share/cds/heapShared.cpp [hunk 1]
```diff
@@ -461,11 +461,13 @@ void HeapShared::archive_objects(ArchiveHeapInfo *heap_info) {
...
-                   UseCompressedOops ? p2i(CompressedOops::begin()) :
-                                       p2i((address)G1CollectedHeap::heap()->reserved().start()),
-                   UseCompressedOops ? p2i(CompressedOops::end()) :
-                                       p2i((address)G1CollectedHeap::heap()->reserved().end()));
+    if (UseCompressedOops || UseG1GC) {
+      log_info(cds)("Heap range = [" PTR_FORMAT " - "  PTR_FORMAT "]",
+                    UseCompressedOops ? p2i(CompressedOops::begin()) :
+                                        p2i((address)G1CollectedHeap::heap()->reserved().start()),
+                    UseCompressedOops ? p2i(CompressedOops::end()) :
+                                        p2i((address)G1CollectedHeap::heap()->reserved().end()));
```
Comment (dholmes-ora) [inline src/hotspot/share/cds/heapShared.cpp:469]: Does this work if you build a minimal VM with only SerialGC? I would expect the build to fail.

File: test/hotspot/jtreg/runtime/cds/appcds/sharedStrings/IncompatibleOptions.java [hunk 2]
```diff
@@ -108,9 +108,9 @@ public static void test(String[] args_ignored) throws Exception {
             testDump(1, "-XX:+UseZGC", "-XX:-UseCompressedOops", null, false);
         }
 
-        // incompatible GCs
-        testDump(2, "-XX:+UseParallelGC", "", GC_WARNING, false);
-        testDump(3, "-XX:+UseSerialGC", "", GC_WARNING, false);
+        // ParallelGC and SerialGC now supports dumping heap objects
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/cds/appcds/sharedStrings/IncompatibleOptions.java:111]: The "now" will be dated very soon. The comment is redundant really.

But do we not need other tests for Serial/Parallel and other flags combinations (or do they not apply)?

**Review by dholmes-ora — COMMENTED** [review]
Generally looks good but two small issues.

Thanks.

File: test/hotspot/jtreg/runtime/cds/appcds/sharedStrings/InvalidFileFormat.java [hunk 3]
```diff
@@ -26,6 +26,7 @@
  * @test
  * @summary Check most common errors in file format
  * @requires vm.cds.write.archived.java.heap
+ * @requires vm.flagless
```
Comment (iklam) [inline test/hotspot/jtreg/runtime/cds/appcds/sharedStrings/InvalidFileFormat.java:29]: I think the reason you needed to add `vm.flagless` is SharedStringsUtils explicitly adds `-XX:UseG1GC` (long time ago this was required to dump the Java heap objects). It's better to remove `-XX:UseG1GC`.

You might need to ensure that all tests that use SharedStringsUtils have `@requires vm.cds.write.archived.java.heap`.

You should also check all other test cases to see if we misuse  `-XX:UseG1GC` for a similar reason.

Comment (matias9927) [reply test/hotspot/jtreg/runtime/cds/appcds/sharedStrings/InvalidFileFormat.java:29]: Looks like all the tests that use `SharedStringsUtil` already have `@requires vm.cds.write.archived.java.heap` so they should be fine. `SharedStringsHumongous` requires G1 so I'll leave that as-is which is also consistent with the Leyden repo.

**Review by iklam — APPROVED** [review]

File: src/hotspot/share/cds/heapShared.cpp (hunk 1 above)
Comment (matias9927) [reply src/hotspot/share/cds/heapShared.cpp:469]: Yes, I verified locally that this works on a minimal build

Comment (dholmes-ora) [reply src/hotspot/share/cds/heapShared.cpp:469]: Doh! Of course it does. The minimal VM doesn't have CDS either. Sorry.

If you build explicitly without G1 then I would expect this to fail.

**Comment by dholmes-ora** [comment]
> I suggest changing the description (per the upstream issue) to be clear this is for Parallel and serial GC only not all non-G1 collectors.

By "description" I meant the JBS issue title: "non-G1" should be "parallel and serial".

File: src/hotspot/share/cds/heapShared.cpp (hunk 1 above)
Comment (dholmes-ora) [reply src/hotspot/share/cds/heapShared.cpp:469]: Not that I think we need to accommodate that. I suspect CDS already has some code dependencies on G1

**Review by dholmes-ora — APPROVED** [review]

File: test/hotspot/jtreg/runtime/cds/appcds/sharedStrings/SharedStringsHumongous.java [hunk 4]
```diff
@@ -27,6 +27,7 @@
  * @summary Use a shared string allocated in a humongous G1 region.
  * @comment -- the following implies that G1 is used (by command-line or by default)
  * @requires vm.cds.write.archived.java.heap
+ * @requires vm.flagless
```
Comment (lmesnik) [inline test/hotspot/jtreg/runtime/cds/appcds/sharedStrings/SharedStringsHumongous.java:30]: Shouldn't 
 @requires vm.gc.G1
works better here if only G1 is supported?

**Review by lmesnik — CHANGES_REQUESTED** [review]

File: test/hotspot/jtreg/runtime/cds/appcds/sharedStrings/SharedStringsHumongous.java (hunk 4 above)
Comment (matias9927) [reply test/hotspot/jtreg/runtime/cds/appcds/sharedStrings/SharedStringsHumongous.java]: This is a direct upstream from Leyden which uses `@requires vm.flagless` although I think you're right.

**Review by lmesnik — APPROVED** [review]
Test changes looks good for me.

**Review by dholmes-ora — APPROVED** [review]

**Comment by matias9927** [comment]
Thanks for the reviews @iklam @lmesnik and @dholmes-ora!
/integrate

**Pull request closed without merging**